Contributing.md

```diff
@@ -0,0 +1,114 @@
+- [.NET Framework Design Guidelines](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/framework-design-guidelines-digest.md) 
+
+There are lots of other useful documents on the [.NET CoreCLR](https://github.com/dotnet/coreclr/tree/master/Documentation) and [.NET Core Framework](https://github.com/dotnet/corefx/tree/master/Documentation) documentation sites which are worth reading, although most experienced C# developers will probably have picked up many of those best-practices by osmosis, particularly around performance and memory management. 
+
+## Source Code Organization
+
+Project Orleans has not religiously followed a "One Class Per File" rule, but instead we have tried to use pragmatic judgment to maximize the change of "code understand-ability" for developers on the team. 
+If lots of small-ish classes share a "common theme" and/or are always dealt with together, then it is OK to place those into one source code file in most cases. 
+See for example the various "log consumer" classes were originally placed in single source file, as they represented a single unit of code comprehension.
+
+As a corollary, it is much easier to find the source code for a class if it is in a file with the same name as the class [similar to Java file naming rules], so there is a tension and value judgment here between code find-ability and minimizing / constraining the number of projects in a solution and files within a project [which both have direct impact on the Visual Studio "Opening" and "Building" times for large projects]. 
+
+Code search tools in VS and ReSharper definitely help here.    
+
+
+## Dependencies and Inter-Project References
+
+One topic that we are very strict about is around dependency references between components and sub-systems.
+
+### Component / Project References
+
+References between projects in a solution must always use "**Project References**" rather than "_DLL References_" to ensure that component build relationships are known to the build tools. 
+
+**Right**:
+
+    <ProjectReference Include="..\Orleans\Orleans.csproj">
+      <Project>{BC1BD60C-E7D8-4452-A21C-290AEC8E2E74}</Project>
+      <Name>Orleans</Name>
+    </ProjectReference>
+
+_Wrong_:
+
+    <Reference Include="Orleans" >
+       <HintPath>..\Orleans\bin\Debug\Orleans.dll</HintPath>
+    </Reference>
+
+In order to help ensure we keep inter-project references clean, then on the build servers [and local `Build.cmd` script] we deliberately use side-by-side input `.\src` and output `.\Binaries` directories rather than the more normal in-place build directory structure (eg. `[PROJ]\bin\Release`) used by VS on local dev machines.
+
+### External Assembly References
+
+We use "**soft**" assembly references (semantic versions) when referencing external assemblies, rather than the "_hard_" (4-digit) fully qualified assembly names that many tools try to sprinkle into our lives at every opportunity.
```

Comment:
The whole point of documenting this is because we **DO** want to "keep enforcing this rule".

On  the PR #1115 thread, we actually found even more examples where the nuget tool's heuristics and decision making was bad or even just plain wrong, so "Trust NuGet to do it right" still remains a fallacy and flawed "plan".